File: lcc/3 ano/2semestre/PC/Praticas/ex1.java
import java.util.*


public class ex1 {



interface Jogo 
{
    Partida participa() throws InterruptedException;
}


interface Partida {
    String adivinha(int n);
}
    class JogoImpl implements Jogo
    {
        PartidaImpl partida = new PartidaImpl();
        public synchronized Partida participa() throws InterruptedException
        {
            PartidaImpl minhaPartida = partida;
            minhaPartida.N_jogadores += 1;
            if (minhaPartida.N_jogadores < 4)
            {
                while (minhaPartida.N_jogadores < 4)
                {
                    wait();
                }    
            }
            else
            {
                partida.start();
                partida = new PartidaImpl();
                notifyAll();
                

            }

            return minhaPartida;
        }

    }

    class PartidaImpl implements Partida
    {
        private int N_jogadores = 0;
        int numero = new Random().nextInt(100) + 1;
        boolean timeout = false;
        boolean ganhou = false;
        int tentativas = 0;

        synchronized void timeout()
        {
            timeout = true;
        }
        private void start()
        {
            new Thread( () -> {
                try
                {
                    Thread.sleep(60000);
                } 
                catch(InterruptedException ignored)
                {}
                timeout();
            }).start();
        }
       public synchronized String adivinha(int n)
        {
            tentativas += 1;
            if (ganhou)
            {
                return "PERDEU";
            }

            if (timeout)
            {
              return "TEMPO";   
            }

            if (tentativas > 100)
            {
                 return "TENTATIVAS";
            }

            if (n == numero)
            {
                return "GANHOU";
                ganhou = true;
            }

            if (n > numero)
            {
                return "MENOR";
            }

            if (n < numero) 
            {
                return "MAIOR";
            }
        }
    }
}

File: lcc/3 ano/2semestre/PC/Praticas/ex1.erl
-module(ex1).
-export([create/0, participa/1, adivinha/2]).

create() -> 
    spawn(fun() -> jogo([]) end).


participa(Jogo) -> 
    Jogo ! {participa, self()},
    receive {partida, Partida} -> Partida end.

adivinha(Partida, N) -> 
    Partida ! {adivinha, N, self()},
    receive {result, Res} -> Res end.

jogo(Jogadores) when length(Jogadores) =:= 4 -> 
    Partida = spawn(fun() -> init_partida() end),
    [Jogador ! {partida, Partida} || Jogador <- Jogadores],
    jogo([]);
jogo(Jogadores) ->
    receive
      {participa, Jogador} -> jogo([Jogador | Jogadores])
    end.

init_partida() -> 
    Self = self(),
    spawn(fun() -> receive after 60000 -> Self ! timeout end end),
    Numero = rand:uniform(100),
    partida(Numero,false, false , 0).

partida(Numero, Timeout, Ganhou, Tentativas) -> 
    receive
    { adivinha, N, Jogador} ->
       Res =
        if
            Ganhou -> "PERDEU";
            Timeout -> "TEMPO";
            Tentativas  >= 100 -> "TENTATIVAS";
            Numero < N -> "MENOR";
            Numero > N -> "MAIOR";
            Numero =:= N -> "GANHOU"
        end,
        Jogador ! {result, Res},
        partida(Numero, Timeout, Ganhou orelse Numero =:= N, Tentativas + 1);
    timeout -> partida(Numero, true, Ganhou, Tentativas)
    end.

    